src/builders/formula_builder.rs

```diff
@@ -0,0 +1,70 @@
+use std::collections::HashMap;
+
+use crate::qdrant::*;
+
+/// Builder for the Formula struct, which represents a scoring formula for points.
+///
+/// The Formula struct is used to define custom scoring expressions and default values.
+pub struct FormulaBuilder {
+    /// The expression that defines how to score points.
+    pub(crate) expression: Option<Expression>,
+    /// Default values to use for undefined variables in the expression.
+    pub(crate) defaults: HashMap<String, Value>,
+}
+
+impl FormulaBuilder {
+    /// Sets the expression for the formula.
+    pub fn new<E: Into<Expression>>(expression: E) -> Self {
+        let new = Self::create_empty();
+        new.expression(expression)
+    }
+
+    pub fn expression<E: Into<Expression>>(self, value: E) -> Self {
+        let mut new = self;
+        new.expression = Some(value.into());
+        new
+    }
+
+    /// Sets all default values for the formula.
+    pub fn defaults(mut self, defaults: HashMap<String, Value>) -> Self {
+        self.defaults = defaults;
+        self
+    }
+
+    /// Adds a single default value to the formula's defaults.
+    pub fn add_default<K: Into<String>, V: Into<Value>>(self, key: K, value: V) -> Self {
+        let mut new = self;
+        new.defaults.insert(key.into(), value.into());
+        new
+    }
+
+    fn build_inner(self) -> Result<Formula, std::convert::Infallible> {
+        Ok(Formula {
+            expression: self.expression,
+            defaults: self.defaults,
+        })
+    }
+
+    fn create_empty() -> Self {
+        Self {
+            expression: None,
+            defaults: Default::default(),
+        }
+    }
```

Comment:
Same here, would it be possible to remove the empty expression and make `expression` non-Option?